src/transformers/models/gptj/modeling_gptj.py

```diff
@@ -0,0 +1,1046 @@
+        
+        if scale_attn is not None:
+            attn_weights = attn_weights / scale_attn
+
+        if attention_mask is not None:
+            # Apply the attention mask
+            attn_weights = attn_weights + attention_mask
+
+        attn_weights = nn.Softmax(dim=-1)(attn_weights)
+        attn_weights = attn_weights.to(value.dtype)
+        attn_weights = attn_pdrop(attn_weights)
+
+        # Mask heads if we want to
+        if head_mask is not None:
+            attn_weights = attn_weights * head_mask
+
+        attn_output = torch.matmul(attn_weights, value)
+
+        return attn_output, attn_weights
+
+class GPTJSelfAttention(nn.Module, GPTJAttentionMixin):
+    def __init__(self, config):
+        super().__init__()
+
+        self.window_size = None
+        max_positions = config.n_positions
+        bias = torch.tril(torch.ones((max_positions, max_positions), dtype=torch.uint8)).view(
+            1, 1, max_positions, max_positions
+        ).bool()
+
+        self.register_buffer("bias", bias,)
+
+        self.register_buffer("masked_bias", torch.tensor(-1e9))
+
+        self.attn_pdrop = nn.Dropout(config.attn_pdrop)
+        self.resid_pdrop = nn.Dropout(config.resid_pdrop)
+
+        self.embed_dim = config.n_embd
+        self.num_attention_heads = config.num_attention_heads
+        self.head_dim = self.embed_dim // self.num_attention_heads
+        if config.jax:
+            self.register_buffer("scale_attn", torch.sqrt(torch.tensor(self.head_dim)))
+        else:
+            self.scale_attn = None
+        if self.head_dim * self.num_attention_heads != self.embed_dim:
+            raise ValueError(
+                f"embed_dim must be divisible by num_attention_heads (got `embed_dim`: {self.embed_dim} and `num_attention_heads`: {self.num_attention_heads})."
+            )
+
+        self.k_proj = nn.Linear(self.embed_dim, self.embed_dim, bias=False)
+        self.v_proj = nn.Linear(self.embed_dim, self.embed_dim, bias=False)
+        self.q_proj = nn.Linear(self.embed_dim, self.embed_dim, bias=False)
+        self.out_proj = nn.Linear(self.embed_dim, self.embed_dim, bias=not config.jax)
+        self.rotary = config.rotary
+        self.rotary_dim = None
+        if config.rotary_dim is not None:
+            self.rotary_dim = config.rotary_dim
+
+    def forward(
+        self,
+        hidden_states,
+        attention_mask=None,
+        layer_past=None,
+        head_mask=None,
+        use_cache=False,
+        output_attentions=False,
+    ):
+
+        query = self.q_proj(hidden_states)
+        key = self.k_proj(hidden_states)
+        value = self.v_proj(hidden_states)
+
+        query = self._split_heads(query, self.num_attention_heads, self.head_dim, self.rotary)
+        key = self._split_heads(key, self.num_attention_heads, self.head_dim, self.rotary)
+        value = self._split_heads(value, self.num_attention_heads, self.head_dim, False)
+
+        if self.rotary:
```

Comment:
GPT-J always uses rotary embeddings, right? In that case, this `if` condition could be removed.

Reply:
GPT-J 6B uses RoPE, but [Mesh Transformer JAX also supports T5 bias and learned absolute](https://github.com/kingoflolz/mesh-transformer-jax/blob/845c69617cb68baf7d752fde09eb2227760cd2b0/device_train.py#L156).

Reply:
I see, thanks. Is there any pre-trained GPT-J model with those embeddings types? If not, for now we could just use RoPE.

Reply:
At the current time, no. However I’m not sure it makes sense to start of with a simpler implementation… my understanding of HF policy is that if we merge a RoPE only version then in the future when we want to use another positional embedding that will be viewed as a “different model.”

Reply:
`self.rotary` is currently not defined